File: apps/ai-studio/backend-go/split_indexes.go
package main

import (
	"errors"
	"strings"
)

func (svc *service) loadSplitMeta() (map[string]any, error) {
	value, err := svc.storageGetByKey(splitMetaKey)
	if err != nil {
		return nil, err
	}
	meta, _ := value.(map[string]any)
	if meta == nil {
		return nil, errors.New("存储未初始化")
	}
	if int(asInt64(meta["schemaVersion"], 0)) != 1 {
		return nil, errors.New("存储索引损坏：meta/index 格式不正确")
	}

	chatsIndex, _ := svc.loadObject(splitChatsIndexKeyGo())
	roleOrder := normalizeStringIDsKeepOrder(firstNonNil(chatsIndex["roleOrder"], meta["roleOrder"]))
	roleFolders := normalizeStringMapGo(firstNonNil(chatsIndex["roleFolders"], meta["roleFolders"]))
	chatIndexByRole := map[string]any{}
	for _, roleID := range roleOrder {
		folder := strings.TrimSpace(roleFolders[roleID])
		if folder == "" {
			continue
		}
		idx, _ := svc.loadObject(splitRoleChatIndexKeyGo(folder))
		if idx == nil {
			idx = asMap(asMap(meta["chatIndexByRole"])[roleID])
		}
		chatIndexByRole[roleID] = normalizeChatIndexSnapshot(idx)
	}

	groupsIndex, _ := svc.loadObject(splitGroupsIndexKeyGo())
	groupOrder := normalizeStringIDsKeepOrder(firstNonNil(groupsIndex["groupOrder"], meta["groupOrder"]))
	groupFolders := normalizeStringMapGo(firstNonNil(groupsIndex["groupFolders"], meta["groupFolders"]))
	chatIndexByGroup := map[string]any{}
	for _, groupID := range groupOrder {
		folder := strings.TrimSpace(groupFolders[groupID])
		if folder == "" {
			continue
		}
		idx, _ := svc.loadObject(splitGroupChatIndexKeyGo(folder))
		if idx == nil {
			idx = asMap(asMap(meta["chatIndexByGroup"])[groupID])
		}
		chatIndexByGroup[groupID] = normalizeChatIndexSnapshot(idx)
	}

	meta["roleOrder"] = roleOrder
	meta["roleFolders"] = roleFolders
	meta["chatIndexByRole"] = chatIndexByRole
	meta["groupOrder"] = groupOrder
	meta["groupFolders"] = groupFolders
	meta["chatIndexByGroup"] = chatIndexByGroup
	return meta, nil
}

func (svc *service) loadProviders() ([]any, error) {
	metaValue, err := svc.storageGetByKey(splitMetaKey)
	if err != nil {
		return nil, err
	}
	meta := asMap(metaValue)
	index, _ := svc.loadObject(splitProvidersIndexKeyGo())
	providerOrder := normalizeStringIDsKeepOrder(index["providerOrder"])
	providerFolders := normalizeStringMapGo(index["providerFolders"])
	if len(providerOrder) == 0 {
		providers := asSlice(asMap(meta["settings"])["providers"])
		if providers == nil {
			return []any{}, nil
		}
		return providers, nil
	}
	providers := make([]any, 0, len(providerOrder))
	for _, providerID := range providerOrder {
		folder := strings.TrimSpace(providerFolders[providerID])
		if folder == "" {
			continue
		}
		provider, err := svc.loadObject(splitProviderKeyGo(folder))
		if err != nil {
			return nil, err
		}
		if provider != nil {
			providers = append(providers, provider)
		}
	}
	return providers, nil
}

func normalizeChatIndexSnapshot(idx map[string]any) map[string]any {
	if idx == nil {
		idx = map[string]any{}
	}
	return map[string]any{
		"activeChatId":  strings.TrimSpace(asString(idx["activeChatId"])),
		"chatIds":       normalizeStringIDsKeepOrder(idx["chatIds"]),
		"chatUpdatedAt": asMapOrNew(idx["chatUpdatedAt"]),
	}
}

func normalizeStringMapGo(raw any) map[string]string {
	src := asMap(raw)
	out := map[string]string{}
	for key, value := range src {
		k := strings.TrimSpace(key)
		v := strings.TrimSpace(asString(value))
		if k != "" && v != "" {
			out[k] = v
		}
	}
	return out
}

func normalizeStringIDsKeepOrder(raw any) []string {
	items := asSlice(raw)
	out := make([]string, 0, len(items))
	seen := map[string]bool{}
	for _, item := range items {
		value := strings.TrimSpace(asString(item))
		if value == "" || seen[value] {
			continue
		}
		seen[value] = true
		out = append(out, value)
	}
	return out
}

func asMapOrNew(raw any) map[string]any {
	value := asMap(raw)
	if value == nil {
		return map[string]any{}
	}
	return value
}

func firstNonNil(items ...any) any {
	for _, item := range items {
		if item != nil {
			return item
		}
	}
	return nil
}

func splitChatsIndexKeyGo() string {
	return "chats/index"
}

func splitRoleChatIndexKeyGo(folder string) string {
	return "chats/" + strings.TrimSpace(folder) + "/index"
}

func splitGroupsIndexKeyGo() string {
	return "groups/index"
}

func splitGroupChatIndexKeyGo(folder string) string {
	return "groups/" + strings.TrimSpace(folder) + "/chats/index"
}

func splitProvidersIndexKeyGo() string {
	return "providers/index"
}

func splitProviderKeyGo(folder string) string {
	return "providers/" + strings.TrimSpace(folder) + "/provider"
}

File: apps/ai-studio/src/storage/splitIndexes.ts
import { now } from '../core/utils'
import { SPLIT_META_KEY, SPLIT_SCHEMA_VERSION } from '../domain/constants'
import { normalizeChatMetas } from '../domain/chatMeta'
import { normalizeSplitMeta } from '../domain/dataNormalizers'
import {
  splitChatsIndexKey,
  splitRoleChatIndexKey,
  splitGroupsIndexKey,
  splitGroupChatIndexKey,
  splitProvidersIndexKey,
  splitProviderKey,
} from '../domain/storageKeys'

export type SplitStorageReader = {
  get: (key: string) => Promise<any>
}

function objectOrEmpty(value: any): Record<string, any> {
  return value && typeof value === 'object' && !Array.isArray(value) ? value : {}
}

function stringList(value: any): string[] {
  return Array.isArray(value) ? value.map((x: any) => String(x || '').trim()).filter(Boolean) : []
}

function stringMap(value: any): Record<string, string> {
  const src = objectOrEmpty(value)
  const out: Record<string, string> = {}
  for (const [key, raw] of Object.entries(src)) {
    const k = String(key || '').trim()
    const v = String(raw || '').trim()
    if (k && v) out[k] = v
  }
  return out
}

function maxUpdatedAt(...items: any[]): number {
  let out = 0
  for (const item of items) {
    const n = Number(item || 0)
    if (Number.isFinite(n) && n > out) out = n
  }
  return out
}

async function readObject(storage: SplitStorageReader, key: string): Promise<Record<string, any>> {
  try {
    const value = await storage.get(key)
    return objectOrEmpty(value)
  } catch (_) {
    return {}
  }
}

export async function loadProvidersFromStorage(storage: SplitStorageReader, metaOverride?: any): Promise<any[]> {
  const index = await readObject(storage, splitProvidersIndexKey())
  const providerOrder = stringList(index.providerOrder)
  const providerFolders = stringMap(index.providerFolders)

  if (!providerOrder.length && metaOverride?.settings && Array.isArray(metaOverride.settings.providers)) {
    return metaOverride.settings.providers.filter((x: any) => x && typeof x === 'object')
  }

  const providers: any[] = []
  for (const providerId of providerOrder) {
    const folder = providerFolders[providerId]
    if (!folder) continue
    const provider = await readObject(storage, splitProviderKey(folder))
    if (!provider || !Object.keys(provider).length) continue
    providers.push(provider)
  }
  return providers
}

export async function loadSplitMetaSnapshot(storage: SplitStorageReader) {
  const raw = await storage.get(SPLIT_META_KEY)
  if (raw == null) return null
  const meta = normalizeSplitMeta(raw)
  if (!meta) throw new Error('存储索引损坏：meta/index 格式不正确')

  const chatsIndex = await readObject(storage, splitChatsIndexKey())
  const roleOrder = stringList(chatsIndex.roleOrder).length ? stringList(chatsIndex.roleOrder) : stringList(meta.roleOrder)
  const roleFolders = Object.keys(stringMap(chatsIndex.roleFolders)).length ? stringMap(chatsIndex.roleFolders) : stringMap(meta.roleFolders)
  const chatIndexByRole: Record<string, any> = {}

  let updatedAt = maxUpdatedAt(meta.updatedAt, chatsIndex.updatedAt)
  for (const roleId of roleOrder) {
    const folder = roleFolders[roleId]
    if (!folder) continue
    const idx = await readObject(storage, splitRoleChatIndexKey(folder))
    chatIndexByRole[roleId] = {
      activeChatId: String(idx.activeChatId || ''),
      chatIds: stringList(idx.chatIds),
      chatUpdatedAt: objectOrEmpty(idx.chatUpdatedAt),
      chatMetas: normalizeChatMetas(idx.chatMetas, idx.chatIds, idx.chatUpdatedAt, '新聊天'),
    }
    updatedAt = maxUpdatedAt(updatedAt, idx.updatedAt)
  }

  const groupsIndex = await readObject(storage, splitGroupsIndexKey())
  const groupOrder = stringList(groupsIndex.groupOrder)
  const groupFolders = stringMap(groupsIndex.groupFolders)
  const chatIndexByGroup: Record<string, any> = {}
  updatedAt = maxUpdatedAt(updatedAt, groupsIndex.updatedAt)

  for (const groupId of groupOrder) {
    const folder = groupFolders[groupId]
    if (!folder) continue
    const idx = await readObject(storage, splitGroupChatIndexKey(folder))
    chatIndexByGroup[groupId] = {
      activeChatId: String(idx.activeChatId || ''),
      chatIds: stringList(idx.chatIds),
      chatUpdatedAt: objectOrEmpty(idx.chatUpdatedAt),
      chatMetas: normalizeChatMetas(idx.chatMetas, idx.chatIds, idx.chatUpdatedAt, '群聊'),
    }
    updatedAt = maxUpdatedAt(updatedAt, idx.updatedAt)
  }

  const providersIndex = await readObject(storage, splitProvidersIndexKey())

  return {
    ...meta,
    schemaVersion: SPLIT_SCHEMA_VERSION,
    updatedAt: updatedAt || now(),
    roleOrder,
    roleFolders,
    chatIndexByRole,
    groupOrder,
    groupFolders,
    chatIndexByGroup,
    providerOrder: stringList(providersIndex.providerOrder),
    providerFolders: stringMap(providersIndex.providerFolders),
  }
}
